e_chil.c

```diff
@@ -340,57 +275,132 @@ static HWCryptoHook_InitInfo hwcrhk_globals = {
+        || !RSA_meth_set_priv_enc(hwcrhk_rsa,
+                                 RSA_meth_get_priv_enc(ossl_rsa_meth))
+        || !RSA_meth_set_priv_dec(hwcrhk_rsa,
+                                 RSA_meth_get_priv_dec(ossl_rsa_meth))
+        || !RSA_meth_set_mod_exp(hwcrhk_rsa, hwcrhk_rsa_mod_exp)
+        || !RSA_meth_set_bn_mod_exp(hwcrhk_rsa, hwcrhk_rsa_bn_mod_exp)
+        || !RSA_meth_set_finish(hwcrhk_rsa, hwcrhk_rsa_finish)) {
+        goto err;
+    }
+#  endif
+
+#  ifndef OPENSSL_NO_DH
+    /* Setup DH Method */
+    hwcrhk_dh = DH_meth_new("CHIL DH method", 0);
+    if (hwcrhk_dh == NULL)
+        goto err;
 
-#ifndef OPENSSL_NO_DH
     /* Much the same for Diffie-Hellman */
-    meth2 = DH_OpenSSL();
-    hwcrhk_dh.generate_key = meth2->generate_key;
-    hwcrhk_dh.compute_key = meth2->compute_key;
-#endif
+    ossl_dh_meth = DH_OpenSSL();
+    if (   !DH_meth_set_generate_key(hwcrhk_dh,
+                                DH_meth_get_generate_key(ossl_dh_meth))
+        || !DH_meth_set_compute_key(hwcrhk_dh,
+                                DH_meth_get_compute_key(ossl_dh_meth))
+        || !DH_meth_set_bn_mod_exp(hwcrhk_dh, hwcrhk_dh_bn_mod_exp)) {
+        goto err;
+    }
+
+#  endif
+
+    if (!ENGINE_set_id(e, engine_hwcrhk_id) ||
+        !ENGINE_set_name(e, engine_hwcrhk_name) ||
+        !ENGINE_set_flags(e, ENGINE_FLAGS_NO_REGISTER_ALL) ||
+#  ifndef OPENSSL_NO_RSA
+        !ENGINE_set_RSA(e, hwcrhk_rsa) ||
+#  endif
+#  ifndef OPENSSL_NO_DH
+        !ENGINE_set_DH(e, hwcrhk_dh) ||
+#  endif
+        !ENGINE_set_RAND(e, &hwcrhk_rand) ||
+        !ENGINE_set_destroy_function(e, hwcrhk_destroy) ||
+        !ENGINE_set_init_function(e, hwcrhk_init) ||
+        !ENGINE_set_finish_function(e, hwcrhk_finish) ||
+        !ENGINE_set_ctrl_function(e, hwcrhk_ctrl) ||
+        !ENGINE_set_load_privkey_function(e, hwcrhk_load_privkey) ||
+        !ENGINE_set_load_pubkey_function(e, hwcrhk_load_pubkey) ||
+        !ENGINE_set_cmd_defns(e, hwcrhk_cmd_defns)) {
+        goto err;
+    }
 
     /* Ensure the hwcrhk error handling is set up */
     ERR_load_HWCRHK_strings();
+
     return 1;
+
+ err:
+    HWCRHKerr(HWCRHK_F_BIND_HELPER, ERR_R_MALLOC_FAILURE);
+
+    CRYPTO_THREAD_lock_free(chil_lock);
+    chil_lock = NULL;
+
+#  ifndef OPENSSL_NO_RSA
+    RSA_meth_free(hwcrhk_rsa);
+    hwcrhk_rsa = NULL;
+#  endif
+#  ifndef OPENSSL_NO_DH
+    DH_meth_free(hwcrhk_dh);
+    hwcrhk_dh = NULL;
+#  endif
+
+    return 0;
+}
+
+#  ifdef OPENSSL_NO_DYNAMIC_ENGINE
+static ENGINE *engine_chil(void)
+{
+    ENGINE *ret = ENGINE_new();
+    if (ret == NULL)
+        return NULL;
+    if (!bind_helper(ret)) {
+        ENGINE_free(ret);
+        return NULL;
+    }
+    return ret;
+}
+
+void engine_load_chil_int(void)
+{
+    /* Copied from eng_[openssl|dyn].c */
+    ENGINE *toadd = engine_chil();
+    if (!toadd)
+        return;
+    ENGINE_add(toadd);
+    ENGINE_free(toadd);
+    ERR_clear_error();
 }
+#  endif
```

Comment:
You can remove the whole section of code from line 380.

Reply:
Removed